worlds/witness/rules.py

```diff
@@ -0,0 +1,168 @@
+            return False
+
+        return True
+
+    def meets_item_requirements(self, panel, world, player, logic, locat):
+        """
+        Checks whether item and panel requirements are met for
+        a panel
+        """
+
+        panel_obj = logic.CHECKS_BY_HEX[panel]
+
+        for option in panel_obj["requirement"]:
+            if len(option) == 0:
+                return True
+
+            valid_option = True
+
+            for item in option:
+                if item == "7 Lasers":
+                    if not self._witness_has_lasers(world, player, 7):
+                        valid_option = False
+                        break
+                elif item == "11 Lasers":
+                    if not self._witness_has_lasers(world, player, 11):
+                        valid_option = False
+                        break
+                elif item in logic.NECESSARY_EVENT_PANELS:
+                    if logic.CHECKS_BY_HEX[item]["checkName"] + " Solved" in locat.EVENT_LOCATION_TABLE:
+                        valid_option = self.has(logic.EVENT_ITEM_NAMES[item], player)
+                    else:
+                        valid_option = self.can_reach(logic.CHECKS_BY_HEX[item]["checkName"], "Location", player)
+                    if not valid_option:
+                        break
+                elif not self.has(item, player):
+                    valid_option = False
+                    break
+
+            if valid_option:
+                return True
+
+        return False
+
+    def _safe_manual_panel_check(self, panel, world, player, logic, locat):
+        """
+        nested can_reach can cause problems, but only if the region being
+        checked is neither of the two original regions from the first
+        can_reach.
+        A nested can_reach is okay here because the only panels this
+        function is called on are panels that exist on either side of all
+        connections they are required for.
+        The spoiler log looks so much nicer this way,
+        it gets rid of a bunch of event items, only leaving a couple. :)
+        """
+        region = logic.CHECKS_BY_HEX[panel]["region"]["name"]
+
+        return (
+                self.meets_item_requirements(panel, world, player, logic, locat)
+                and self.can_reach(region, "Region", player)
+        )
+
+    def can_solve_panels(self, panel_hex_to_solve_set, world, player, logic, locat):
+        """
+        Checks whether a set of panels can be solved.
+        """
+
+        for option in panel_hex_to_solve_set:
+            if len(option) == 0:
+                return True
+
+            valid_option = True
+
+            for panel in option:
+                if not self._can_solve_panel(panel, world, player, logic, locat):
+                    valid_option = False
+                    break
+
+            if valid_option:
+                return True
+        return False
```

Comment:
as per https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/api.md#logic-mixin all methods you add to state through LogicMixin should be prefixed by `_witness_`. You can use pycharm's rightclick -> refactor -> rename to do that.

Basically whatever you add in a class that inherits from LogicMixin is added to the "global" State class. To avoid naming collisions with other worlds, we add a world-specific prefix.